Df 2227 - dist directory

anselmbradford opened the pull request:
Moves project architecture to having a /src/ and /dist/ directory

## Additions

- Added `TEST.md` readme file for testing instructions.
- Added `grunt clean` and `grunt copy` tasks, and added `grunt clean` to `setup.sh`.

## Removals

- Removed `_tests/browser_tests/README.md`, `_tests/macro_testing/README.md`, `_tests/processor_tests/README.md`.

## Changes

- Changed project architecture to having `/src/` and `/dist/` directory.
- Changed `/_tests/` directory name to `/test/`.
- Changed `/_tests/macro_testing` directory name to `/test/macro_tests`.
- Moved `browserify-shims.js` to `/config/` directory.

## Testing

1. Run `./setup.sh`
2. Stop sheer.
3. `cd dist`
4. Start sheer.
5. ???
6. Profit. Site should function the same and tests should run.

## Review

- @dpford 
- @kurtw 
- @jimmynotjim 
- @KimberlyMunoz 
- @sebworks 
- @rosskarchner 

## Notes

- The topdoc task is picking up CSS from the /dist/ directory and generating docs for slick and pdfreactor-fonts. Since we don't use that task anyway I just put a TODO in the Gruntfile for it.
- @dpford please check that links.sh was updated correctly.
- @dpford @kurtw the processor tests are failing, but I'm not sure if they were failing before. Says, e.g.:
  ```
ERROR: test_views (tests.test_wordpress_view_processor.WordpressViewProcessorTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/bradforda/homebrew/lib/python2.7/site-packages/mock.py", line 1201, in patched
    return func(*args, **keywargs)
  File "/Users/bradforda/Projects/cfgov-refresh/test/processor_tests/tests/test_wordpress_view_processor.py", line 37, in test_views
    custom_fields = document['custom_fields']
KeyError: 'custom_fields'
```
- @sebworks please check that crawler code in `/src/external-site/_vars-external-site.html` was updated correctly.
- Fixes https://github.com/cfpb/cfgov-refresh/issues/750. Full changes are in that issue.

anselmbradford commented on src/external-site/_vars-external-site.html:
```diff
@@ -6,7 +6,7 @@
 
     // Find the external-urls in your local cfgov-refesh repo.
-    find . -name '*.html' ! -path './src/*' ! -path './node_modules/*' ! -path './_tests/*' \
+    find . -name '*.html' ! -path './*' ! -path '../node_modules/*' ! -path '../test/*' \
```
I don't think I updated this right. What's this path relative to and how do I test?

sebworks replied:
It should run in the root of your cfgov-refresh directory. Best way to test it is to run the commands before/after the changes and compare.

anselmbradford replied:
Cool. Just noticing this is wrapped in a code comment. So you'd copy and paste this into a script in root and run it? Is that right? How often does it need to be run and what's the relationship (if any) to `./links.sh`?

anselmbradford replied:
Updated paths in https://github.com/cfpb/cfgov-refresh/commit/cca1733c3ae120ee9ebb47d8780932f2d8f67b11

sebworks replied:
That is correct but this code can actually be removed at this point since we have the `.links.sh`. The code in `.link.sh` is accomplishing the same thing and is running as a Cron job of some sort.

anselmbradford replied:
Removed in https://github.com/cfpb/cfgov-refresh/commit/0721281b3f935582117a39e179c86fcd4dd61223

rosskarchner commented:
I've been eyeing this setup:

https://lincolnloop.com/blog/simplifying-your-django-frontend-tasks-grunt/

Is there anything stopping us from having  'grunt watch'  available?

anselmbradford commented:
> Is there anything stopping us from having 'grunt watch' available?

We have `grunt watch` [here](https://github.com/cfpb/cfgov-refresh/blob/flapjack/Gruntfile.js#L307-L324) or did you mean something else?

rosskarchner commented:
ah, then I just missed it

anselmbradford commented:
Also, thanks for commenting! Totally forgot to add you to the review list... (stealth edited)

sebworks commented:
@anselmbradford , why aren't we using the latest for all grunt dependencies?

anselmbradford commented:
> why aren't we using the latest for all grunt dependencies?

@sebworks dependencies should be updated individually in their own individual PRs in order to isolate any breaking changes. grunt-contrib-copy and grunt-contrib-clean are both new here so are set on latest. AFAIK dev-dependencies that do simple tasks (such as copying and removing directories) can be versioned against latest stable, but others should be pegged to a version to avoid breaking surprises.

sebworks commented:
Will we lose the ability to view history for all the files now that they are moving?

anselmbradford commented:
> Will we lose the ability to view history for all the files now that they are moving?

I'm not sure, but I don't think so. Other than the two test readmes, the file lists shows moved files, not removed and added files, so I think git's smart enough to track them.

sebworks commented:
:+1:

dpford commented:
![upvote](http://replygif.net/i/1118.gif)

anselmbradford commented:
I found that `./setup.sh

anselmbradford commented:
While doing final testing I found that `./setup.sh` and `grunt clean` followed by `grunt build` did not produce the same result, which shouldn't be the case. It was because `./setup.sh` runs `grunt vendor` separately. This heralds from when `grunt vendor` downloaded the bower dependencies, so we didn't include it in the build step. This doesn't happen anymore, so I removed `grunt vendor` from `setup.sh` and added it to `grunt build`. Since this came up after squashing and this PR touches a lot, I'm going to merge those changes in. Hope that's okay.

jimmynotjim commented:
Tested w/ latest update and still working as expected :+1:
